syntaxes/kotlin.tmLanguage.json

```diff
@@ -368,10 +391,41 @@
                 "5": {
                     "name": "entity.name.function.declaration.kotlin"
                 }
+            },
+            "end": "\\)",
+            "endCaptures": {
+                "1": {
+                    "name": "keyword.operator.assignment.type.kotlin"
+                }
+            },
+            "patterns": [
+                {
+                    "include": "#parameter-declaration"
+                }
+            ]
+        },
+        "parameter-declaration": {
+            "match": "\\b(\\w+)\\s*(:)\\s*(\\w+)(\\?)?(,)?",
+            "captures": {
+                "1": {
+                    "name": "variable.parameter.kotlin"
+                },
+                "2": {
+                    "name": "keyword.operator.assignment.type.kotlin"
+                },
+                "3": {
+                    "name": "entity.name.type.kotlin"
+                },
+                "4": {
+                    "name": "keyword.operator.optional"
+                },
+                "5": {
+                    "name": "punctuation.separator.delimiter.kotlin"
+                }
             }
         },
         "variable-declaration": {
-            "match": "\\b(val|var)\\b\\s*(?<GROUP><([^<>]|\\g<GROUP>)+>)?",
+            "match": "\\b(var)\\b\\s*(?<GROUP><([^<>]|\\g<GROUP>)+>)?",
```

Comment:
The separation of `var` and `val` is interesting as it lets us highlight constants differently.

On the other hand, I worry that this duplication will increase the maintainance burden. In general, I think "being liberal" about allowing the same syntactic constructs for `var` and `val` should be fine for syntax highlighting.

Curious to know how other languages handle this.

Reply:
To be honest, I'm not sure where I got the idea to create a separate scope for val. I have checked how the TypeScript grammar handles this, and it combines let, const and var.

C# uses readonly scope, but only for the readonly keyword https://github.com/microsoft/vscode/blob/83127004a87124b1b8b976b9f9022da41814dfd0/extensions/csharp/syntaxes/csharp.tmLanguage.json#L3034

Reply:
I have searched a bit more, and I added the readonly distinction because of the "Semantic Highlight Guide" documentations for VSCode mentions it.
See: https://code.visualstudio.com/api/language-extensions/semantic-highlight-guide
![image](https://github.com/fwcd/vscode-kotlin/assets/25014042/5b38e4f0-b5c3-46df-bf8d-fa19a19145c2)
However, it does not look like it is common to be used for the const/val keywords, only for the readonly keyword.
I can revert the change, it you prefer.

I have also made some changes to the definitions  for val/var to be more in line with how the java grammar handles local var.

I may be a bit biased when I'm using the Java grammar for guidance, but I'm only using Kotlin in conjunction with Java, so that's what i know.